nautobot/extras/groups.py

```diff
@@ -0,0 +1,182 @@
+    form_class = get_form_for_model(model)
+
+    group_map = type(
+        str("%sDynamicGroupMap" % model._meta.object_name),
+        (BaseDynamicGroupMap,),
+        {
+            "model": model,
+            "filterset_class": filterset_class,
+            "filterform_class": filterform_class,
+            "form_class": form_class,
+        },
+    )
+
+    return group_map
+
+
+class BaseDynamicGroupMap:
+    """
+    Dynamic Group mapping used to generate mappings for each model class.
+
+    This class itself should not be invoked directly as the class variables will
+    not be populated and most class methods will fail.
+    """
+
+    model = None
+    form_class = None
+    filterset_class = None
+    filterform = None
+
+    # This is used as a `startswith` check on field names, so these can be
+    # explicit fields or just substrings.
+    #
+    # Currently this means skipping "search", custom fields, and custom relationships.
+    #
+    # FIXME(jathan): As one example, `DeviceFilterSet.q` filter searches in `comments`. The issue
+    # really being that this field renders as a textarea and it's not cute in the UI. Might be able
+    # to dynamically change the widget if we decide we do want to support this field.
+    #
+    # Type: tuple
+    exclude_filter_fields = ("q", "cr", "cf", "comments")  # Must be a tuple
+
+    @classproperty
+    def base_url(cls):
+        if cls.model is None:
+            return
+        route_name = get_route_for_model(cls.model, "list")
+        return reverse(route_name)
+
+    @classmethod
+    def fields(cls):
+        """Return all FilterForm fields in a dictionary."""
+
+        # Get model form and fields
+        modelform = cls.form_class()
+        modelform_fields = modelform.fields
+
+        # Get filter form and fields
+        filterform = cls.filterform_class()
+        filter_fields = filterform.fields
+
+        # Get filterset and fields
+        filterset = cls.filterset_class()
+        filterset_fields = filterset.filters
+
+        # Get dynamic group filter field mappings (if any)
+        dynamic_group_filter_fields = getattr(cls.model, "dynamic_group_filter_fields", {})
+
+        # Model form fields that aren't on the filter form
+        missing_fields = set(modelform_fields).difference(filter_fields)
+
+        # Try a few ways to see if a missing field can be added to the filter fields.
+        for missing_field in missing_fields:
+            # Skip excluded fields
+            if missing_field.startswith(cls.exclude_filter_fields):
+                logger.debug("Skipping excluded filter field: %s", missing_field)
+                continue
+
+            # In some cases, fields exist in the filterset AND by another name (e.g. `cluster` ->
+            # `cluster_id` yet are ommited from the filter form. We only want to
+            # add them if-and-only-if they aren't already in `filter_fields`.
```

Comment:
This comment isn't super clear to me - it reads like the idea that `cluster` and `cluster_id` are both filterset fields and we don't want to duplicate them? But looking at VirtualMachineFilterSet I only see `cluster_id` - so what exactly is going on here?

Reply:
Here goes:

- They are both filterset fields
- On `VirtualMachineFilterset`, `cluster` is listed in `Meta.fields`, so it's *default behavior* is inhertied
- On `VirtualMachineFilterForm`, `cluster` is not included at all.
  - `cluster_id` shadows `cluster` in that it behaves differently. It's explicitly a `pk` only field, and not a "related" field like `cluster`. 
- So since `cluster` is already represented in the UI form as `cluster_id`, we DO NOT want `cluster` to be magically added, because then it results in TWO fields called "Cluster" in the UI filter form, each of which do the same thing but in a different way. 

🤯